File: LeetCode/0433. Minimum Genetic Mutation/solution.cpp
/**
 * 14 / 14 test cases passed.
 * Runtime: 0 ms
 * Memory Usage: 6.5 MB 
 */
class Solution {
public:
    int min_steps = INT_MAX;
    unordered_set<string> used;
    int minMutation(string start, string end, vector<string>& bank) {
        dfs(0, start, end, bank);
        return min_steps == INT_MAX ? -1 : min_steps;
    }
    void dfs(int step, string curr, string end, vector<string>& bank) {
        if (curr == end) {
            min_steps = min(min_steps, step);
        }
        for (auto &gene : bank) {
            int diff = 0;
            for (int i = 0; i < gene.size() && diff <= 1; ++ i) {
                if (gene[i] != curr[i]) ++ diff;
            }
            if (diff == 1 && !used.count(gene)) {
                used.insert(gene);
                dfs(step + 1, gene, end, bank);
                used.erase(gene);
            }
        }
    }
};

File: LeetCode/0433. Minimum Genetic Mutation/solution.py
"""
14 / 14 test cases passed.
Runtime: 40 ms
Memory Usage: 15.1 MB
"""
class Solution:
    def minMutation(self, start: str, end: str, bank: List[str]) -> int:
        used, bank = set(), set(bank)
        if end not in bank: return -1
        # current gene, steps
        que = collections.deque([(start, 0)])
        while que:
            cur, step = que.popleft()
            if cur == end:
                return step
            for gene in bank:
                diff = 0
                for i in range(len(gene)):
                    if cur[i] != gene[i]:
                        diff += 1
                if diff == 1 and gene not in used:
                    used.add(gene)
                    que.append((gene, step + 1))
        return -1
